src/app/keyboard/components/virtual-keyboard/virtual-keyboard.component.ts

```diff
@@ -1,31 +1,89 @@
-import { Component, Input } from '@angular/core'
-import { KeyBinding } from '../../models/key-binding'
+import { Component, Input, OnChanges, SimpleChanges } from '@angular/core'
 import { KeyBindingMatcher } from '../../abstractions/key-binding-matcher.service'
-import { KeyEventType } from '../../value-objects/key-event-type'
+import { Logger } from '../../../core/abstractions/logger.service'
+import { StyledKeyBinding } from '../../value-objects/styled-key-binding'
+import { KeyAliasService } from '../../abstractions/key-alias-service'
+import { PhysicalKeyboardLayoutService } from '../../services/physical-keyboard-layout.service'
+import { KeyboardLayout, KeyboardLayoutKey } from '../../value-objects/keyboard-layout'
 
 @Component({
   selector: 'sofie-virtual-keyboard',
   templateUrl: './virtual-keyboard.component.html',
   styleUrls: ['./virtual-keyboard.component.scss'],
 })
-export class VirtualKeyboardComponent {
+export class VirtualKeyboardComponent implements OnChanges {
   @Input()
-  public keyBindings: KeyBinding[]
+  public keyBindings: StyledKeyBinding[]
 
   @Input()
   public keystrokes: string[]
 
-  constructor(private readonly keyBindingMatcher: KeyBindingMatcher) {}
+  public keyboardLayoutMap: KeyboardLayoutMap = new Map()
+  public physicalKeyboardLayout: KeyboardLayout
 
-  public displayKeyBinding(keyBinding: KeyBinding): string {
-    return keyBinding.keys.map(key => this.displayKey(key)).join(' + ')
+  private keyBindingsFilteredByModifiers: StyledKeyBinding[] = []
+  private currentModifierKeystrokes: string[] = []
+  private readonly logger: Logger
+
+  constructor(
+    private readonly keyBindingMatcher: KeyBindingMatcher,
+    private readonly keyAliasService: KeyAliasService,
+    logger: Logger
+  ) {
+    this.logger = logger.tag('VirtualKeyboardComponent')
+    this.updatePhysicalKeyboardLayout()
+    navigator.keyboard
+      ?.getLayoutMap()
+      .then(keyboardLayoutMap => (this.keyboardLayoutMap = new Map(keyboardLayoutMap)))
+      .catch(error => this.logger.data(error).warn('Failed getting keyboard layout.'))
+      .finally(() => this.updatePhysicalKeyboardLayout())
+  }
+
+  private updatePhysicalKeyboardLayout(): void {
+    const physicalKeyboardLayoutService: PhysicalKeyboardLayoutService = new PhysicalKeyboardLayoutService(this.keyboardLayoutMap)
```

Comment:
Why is this not dependency injected?

Reply:
I assume it's because you parse in a constructor paramter.
The only thing you call in the constructor is `registerCustomKeyboardLayoutMapEntries()`.
You could just parse the parameter to `createIso102KeyboardLayout` and call the method as the first thing in there.

Sidenote: You have been making abstractions for everything else. Why not for the KeyboardLayoutService as well?
Then you can just call the method `createKeyboardLayout()` and have an `iso102KeyboardLayoutService` implementation. Then this component wouldn't even know that it's using a iso102 keyboard.

Reply:
PhysicalKeyboardLayoutService abstraction is added with the methods `registerKeyboardKeyLabels` and `getPhysicalKeyboardLayout` and DI is used.